chore: use pnpm to manage dependencies

jacobwgillespie opened the pull request:
## Problem

Followup from #12520 (see there for more context), this PR replaces Yarn 1 with pnpm. Yarn 1 is semi-unmaintained and has caused issues in `docker build`.

## Changes

This PR switches to use `pnpm` instead of `yarn`. The highlights are:

* All yarn commands are updated to their pnpm equivalents
* GitHub Actions workflows and the Dockerfile use pnpm now
* Some additional dependencies have been added to `package.json` if they previously were only available due to hoisting.

## How did you test this code?

Installed locally + ran `docker build`. Will need CI to run the tests.

jacobwgillespie committed:
chore: use pnpm to manage dependencies

jacobwgillespie committed:
Fix CI errors

jacobwgillespie committed:
Don't report Docker image size for external PRs

jacobwgillespie committed:
Fix pnpm-lock.yaml formatting

jacobwgillespie committed:
Fix module versions

jacobwgillespie committed:
Ignore pnpm-lock.yaml

jacobwgillespie committed:
Upgrade Cypress action for pnpm support

jacobwgillespie committed:
Set up node and pnpm before Cypress

Twixes reviewed (commented):
_Looks_ good to me, but TypeScript is acting up, and I don't see why since nothing should be different with regard to typing. 🤔 Also something fishy going on with plugin server tests.

jacobwgillespie committed:
Fix typescript issues

jacobwgillespie commented on frontend/src/scenes/insights/filters/ActionFilter/ActionFilter.tsx:
```diff
@@ -76,159 +76,154 @@ export interface ActionFilterProps {
-            addFilterDefaultOptions,
-        })
-        const { reportFunnelStepReordered } = useActions(eventUsageLogic)
+export const ActionFilter = React.forwardRef<HTMLDivElement, ActionFilterProps>(function ActionFilter(
```
That's for the ESLint rule `react/display-name` - with the arrow function inside `forwardRef`, the wrapped component is unnamed.

Twixes replied:
Oh interesting, I thought we were already compliant with this rule. Looks like it's just been a while since this file was last edited and only now did ESLint catch this.

jacobwgillespie committed:
Include patches directory in Dockerfile

jacobwgillespie commented:
> TypeScript is acting up, and I don't see why since nothing should be different with regard to typing

Yeah... So this is the fun of migrating from Yarn 1 to pnpm (or a more strict Yarn) - Yarn 1 hoisted all modules to the root of `node_modules`, so you could potentially import a sub-dependency in your app without specifying it in package.json, and the behavior of tools that inspect `node_modules` can change. For instance, `@types/*` packages would be hoisted to the top, ESLint plugins would be hoisted, etc.

With pnpm or a more correct Yarn version, you can only import packages specified in `package.json`. This is way safer and more predictable, but when migrating you discover all the places that previously were accidentally working due to hoisting. 🙂 

I'll keep watching the build, I think I'm getting closer to fixing all the TS issues.

Twixes commented on patches/kea-forms@3.0.1.patch:
```diff
@@ -0,0 +1,37 @@
+diff --git a/lib/types.d.ts b/lib/types.d.ts
```
CC @mariusandra

jacobwgillespie replied:
Ah, cool, yeah an upstream fix would be even better. 🙂 

The issue here is that `kea` in 3.0.0-alpha accepted just two generic arguments for `SelectorDefinition`, but in the 3.x.x released version requires three. Looks like `kea-forms` assumes the alpha version's types.

mariusandra replied:
Thanks for finding this! I released version 3.0.2 of kea-forms with the fix.

Twixes reviewed (commented):
This is looking fantastic, well done! Ran this locally (`bin/start`) and everything's working.
The only remaining issue it those few failing Cypress tests – though I ran them on a clean setup locally with fully green results, so this might be some flakiness. Also not sure why Jest tests haven't ran in CI. Once the checks are green, we can merge this.

jacobwgillespie commented:
> this might be some flakiness. Also not sure why Jest tests haven't ran in CI. Once the checks are green, we can merge this.

Yeah, having done this before, the flakiness may very well be related to some module reshuffling, and after merge / deploy it's quite possible that there might be a followup issue to fix for the release. 🙂 Whenever this is merged, it would be good to watch for any errors, etc.

jacobwgillespie committed:
Fix Jest tests in CI

jacobwgillespie committed:
Merge branch 'master' into pnpm

jacobwgillespie commented on .github/workflows/ci-frontend.yml:
```diff
@@ -56,21 +52,22 @@ jobs:
+              # Take the 3rd line of the output (the first two are pnpm non-sense)
               # Split the test into 3 parts. To increase the number split change the denominator in `length / 3`
-              run: echo "test-chunks=$(yarn test --listTests --json | sed -n 3p | jq -cM '[_nwise(length / 3 | ceil)]')" >> $GITHUB_OUTPUT
+              run: echo "test-chunks=$(pnpm test -- --listTests --json | sed -n 3p | jq -cM '[_nwise(length / 3 | ceil)]')" >> $GITHUB_OUTPUT
```
Updated 👍 

> Allow edits and access to secrets by maintainers

Apparently GitHub doesn't support this if the repo and the fork are both located in organization accounts rather than personal accounts. 🤷 https://github.com/orgs/community/discussions/5634

Twixes replied:
Oh, that's very annoying. Now there are some "Jest encountered an unexpected token" problems in Frontend CI checks, and unfortunately I'm not able to push a fix either.

jacobwgillespie committed:
Update lockfile

mariusandra reviewed (commented):
This looks good to me! The last changes I think we need are:
- kea-forms can now be upgraded to a version that fixes the bug
- multiple RUN commands in production.Dockerfile should be combined to one, in two places

Other than that, if there's no easy way to get CI working, we're just going to merge it in, and quickly fix things if any remaining issues arise. I think the stakes are low enough that we could get away with it.

jacobwgillespie committed:
Merge branch 'master' into pnpm

jacobwgillespie committed:
Update lockfile

jacobwgillespie committed:
Clean up Dockerfile

jacobwgillespie commented:
Okay, I've updated with changes from `master` and cleaned up the Dockerfile. I still need to fix Jest, which is currently not transpiling tests correctly (it still has `import` / `export`).

mariusandra commented:
We recently upgraded Jest and are running it now through `esbuild`, so that problem might have solved itself.

I'd be happy to merge it in, though we decided to wait until next week as we're just about to release a new version of PostHog... and just in case it's best to wait for after that :).

jacobwgillespie commented:
Closing in favor of #13190 🚀

charlescook-ph commented:
Thank you so much @jacobwgillespie! Check your email for a small gift from us :)

jacobwgillespie commented:
Hey, thank you! 🙌
